### ISSUE TITLE
Crash at launch on macOS 11 ARM64 - setupLogBuffer

### ISSUE THREAD

**Issue by macmade** [microsoft/appcenter-sdk-apple#2511]
### **Description**

We are getting reports of users experiencing crashes at launch on macOS 11 on ARM64.  
The crash appears to be systematic.

```
Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000000000
Exception Note:        EXC_CORPSE_NOTIFY
```

  - Users on macOS 12 are not affected.
  - Users on macOS 11 Intel are not affected.

The crash happens in `[MSACCrashes setupLogBuffer]` line `1167`, called when `MSACCrashes` is initialized.

Looking at the source code for this line, I can see:

      /*
       * Some explanation into what actually happens, courtesy of Gwynne: "Passing nil does not initialize anything to nil here, what
       * actually happens is an exploit of the Objective-C send-to-nil-returns-zero rule, so that the effective initialization becomes
       * `buffer(&(0)[0], &(0)[0])`, and since `NULL` is zero, `[0]` is equivalent to a direct dereference, and `&(*(NULL))` cancels out to
       * just `NULL`, it becomes `buffer(nullptr, nullptr)`, which is a no-op because the initializer code loops as `while(begin != end)`,
       * so the `nil` pointer is never dereferenced."
       */
      msACCrashesLogBuffer[i] = MSACCrashesBufferedLog(path, nil);

`MSACCrashesBufferedLog` does:

     MSACCrashesBufferedLog(NSString *path, NSData *data)
      : bufferPath(path.UTF8String),
        buffer(&reinterpret_cast<const char *>(data.bytes)[0], &reinterpret_cast<const char *>(data.bytes)[data.length]) {}

Since the crash is a segmentation fault on a `NULL` pointer, I'm quite confident that `&reinterpret_cast<const char *>(data.bytes)[0]` is causing the issue.  
Maybe undefined behavior?

Here's the symbolicated stack trace:

```
0x00000001083b4cbc -[MSACCrashes setupLogBuffer] (in DevKit) (MSACCrashes.mm:1167)
0x00000001083ae390 -[MSACCrashes init] (in DevKit) (MSACCrashes.mm:0)
0x00000001083aed74 __29+[MSACCrashes sharedInstance]_block_invoke (in DevKit) (MSACCrashes.mm:481)
0x000000018917fec0 _dispatch_client_callout + 20
0x0000000189181758 _dispatch_once_callout + 32
0x00000001083aed38 +[MSACCrashes sharedInstance] (in DevKit) (MSACCrashes.mm:479)
0x00000001083add10 +[MSACCrashes setDelegate:] (in DevKit) (MSACCrashes.mm:0)
0x0000000108360660 specialized CrashReportWindowController.init(window:) (in DevKit) (CrashReportWindowController.swift:53)
0x000000010835fbfc @objc CrashReportWindowController.init(window:) (in DevKit) (<compiler-generated>:49)
0x000000010835f678 one-time initialization function for shared (in DevKit) (CrashReportWindowController.swift:18)
0x000000018917fec0 _dispatch_client_callout + 20
0x0000000189181758 _dispatch_once_callout + 32
0x000000019537db40 swift_once + 48
0x0000000108353b8c AppCenterIntegration.setup(uuid:secret:) (in DevKit) (AppCenter.swift:154)
0x0000000108353e38 @objc AppCenterIntegration.setup(uuid:secret:) (in DevKit) (<compiler-generated>:0)
0x00000001069b9700 -[ApplicationDelegate applicationDidFinishLaunching:] (in iMazing) (ApplicationDelegate.m:243)
0x00000001894232f8 __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 28
0x00000001894c523c ___CFXRegistrationPost_block_invoke + 52
0x00000001894c51a8 _CFXRegistrationPost + 440
0x00000001893f1dc8 _CFXNotificationPost + 720
0x000000018a17c3c0 -[NSNotificationCenter postNotificationName:object:userInfo:] + 64
0x000000018bc27168 -[NSApplication _postDidFinishNotification] + 340
0x000000018bc0ecd8 -[NSApplication finishLaunching] + 2264
0x000000018bc0e160 -[NSApplication run] + 292
0x000000018bbdfcbc NSApplicationMain + 1064
0x000000018934d430 start + 4//
```

### **Repro Steps**

Please list the steps used to reproduce your issue.

1. Run the app on macOS 11 - ARM64

### **Details**

1. Which SDK version are you using?
    - 5.0.0
2. Which OS version did you experience the issue on?
    - macOS 11 ARM64
3. Which Xcode version did you build the app with?
    - Xcode 15.0
5. Which Cocoapods version are you using (run `pod --version`)?
    - 1.12.0
6. What device version did you see this error on?  Were you using an emulator or a physical device?
    - Physical ARM64 device
7. What language are you using?
    - [x] Objective C
    - [x] Swift
8. What third party libraries are you using?
9. Please enable verbose logging for your app using `MSAppCenter.setLogLevel(.verbose)` before your call to `MSAppCenter.start(...)` for Swift, or `[MSAppCenter setLogLevel:MSLogLevelVerbose]` before `[MSAppCenter start: ...]` for Objective C and include the logs here:

**Comment by macmade** [comment]
IMHO the code in `MSACCrashesBufferedLog` should simply check if the `data` object is `nil` and initialize explicitely `buffer` to `nullptr` instead of relying on compiler tricks to save a few characters of code.

**Comment by DmitriyKirakosyan** [comment]
Hi @macmade , thanks for reaching out! Please check this comment https://github.com/microsoft/appcenter-sdk-apple/issues/2497#issuecomment-1753144003. If you have the opportunity, please try Xcode 15.1 to see if it fixes the issue.

**Comment by macmade** [comment]
The crash can be fixed as well in Xcode 15.0 by forcing the old linker: `-Wl -ld_classic`  
So yeah, will be fixed by 15.1. Closing.
